processor/batchprocessor/splitmetrics.go

```diff
@@ -0,0 +1,65 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package batchprocessor
+
+import (
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+// splitMetrics removes metrics from the input data and returns a new data of the specified size.
+func splitMetrics(size int, toSplit pdata.Metrics) pdata.Metrics {
+	if toSplit.MetricCount() <= size {
+		return toSplit
+	}
+	copiedMetrics := 0
+	result := pdata.NewMetrics()
+	rms := toSplit.ResourceMetrics()
+	for i := rms.Len() - 1; i >= 0; i-- {
+		rm := rms.At(i)
+		destRs := pdata.NewResourceMetrics()
+		rm.Resource().CopyTo(destRs.Resource())
+		result.ResourceMetrics().Append(destRs)
+
+		for j := rm.InstrumentationLibraryMetrics().Len() - 1; j >= 0; j-- {
+			instMetrics := rm.InstrumentationLibraryMetrics().At(j)
+			destInstMetrics := pdata.NewInstrumentationLibraryMetrics()
+			destRs.InstrumentationLibraryMetrics().Append(destInstMetrics)
```

Comment:
Same trick here, it is better to resize the slice than allocating + appending, each allocating +append result in 1 extra allocation for sure + a maybe one when it needs to resize the internal slice.

Reply:
Done in 711767a